torch/_decomp/decompositions.py

```diff
@@ -20,38 +22,36 @@ class Reduction(Enum):
     SUM = 2
 
 
-def cast_for_opmath(f):
+# This wraps a decomposition and performs various type promotion logic within it, depending on the strategy provided
+# We're currently re-using ELEMENTWISE_TYPE_PROMOTION_KIND, although some of the usages are on non-elementwise ops
+# Will need to validate the non-elementwise uses
+def type_casts(f: Callable, type_promotion: refs.ELEMENTWISE_TYPE_PROMOTION_KIND):
     @functools.wraps(f)
     def inner(*args, **kwargs):
-        orig_prec = None
+        flat_args = [x for x in tree_flatten((args, kwargs))[0] if isinstance(x, Tensor)]
+        computation_dtype, result_dtype = refs._elementwise_dtypes(*flat_args, type_promotion=type_promotion)
 
         # TODO: pretty sure this is not quite right
         def increase_prec(x):
-            if isinstance(x, Tensor) and x.dtype in (torch.float16, torch.bfloat16):
-                nonlocal orig_prec
-                if orig_prec is None:
-                    orig_prec = x.dtype
-                else:
-                    assert orig_prec == x.dtype
-                return x.to(torch.float32)
+            if isinstance(x, Tensor):
+                return x.to(computation_dtype)
```

Comment:
This would mask possible type errors for ops that don't support inputs of the different type?

Reply:
yes, for now. I was thinking another wrapper on top of it for situations like that (or for situations where one tensor needs to be say, an integer dtype).

Any thoughts?